src/aleph/web/controllers/storage.py

```diff
@@ -114,59 +117,168 @@ class StorageMetadata(pydantic.BaseModel):
+        hash_sha256 = hashlib.sha256()
+
+        while total_read < self.max_size:
+            chunk = await self.file_field.read_chunk(chunk_size)
+            if not chunk:
+                break
+            total_read += len(chunk)
+            if total_read > self.max_size:
+                raise web.HTTPRequestEntityTooLarge(
+                    reason="File size exceeds the maximum limit.",
+                    max_size=self.max_size,
+                    actual_size=total_read,
+                )
+            self._temp_file.write(chunk)
+            self._file_content.extend(chunk)
+            hash_sha256.update(chunk)
+        self._hash = hash_sha256.hexdigest()
+        self._temp_file.seek(0)
+
+    @property
+    def size(self) -> int:
+        return os.path.getsize(self._temp_file.name)
 
     @property
     def content(self) -> bytes:
-        # Only read the stream once
-        if self._content is None:
-            self.file_field.file.seek(0)
-            self._content = self.file_field.file.read(self.size)
+        return bytes(self._file_content)
 
-        return self._content
+    @property
+    def file(self) -> str:
+        return self._temp_file.name
+
+    def get_hash(self) -> str:
+        if self._hash is None:
+            raise ValueError("Hash has not been computed yet")
+        return self._hash
+
+
+class RawUploadedFile(UploadedFile):
+    def __init__(self, request: web.Request, max_size: int):
+        self.request = request
+        self.max_size = max_size
+        self._temp_file = tempfile.NamedTemporaryFile(delete=False)
+        self._hasher = hashlib.sha256()
+        self._size = 0
+        self._hash = None
+
+    async def read_and_validate(self):
+        async for chunk in self.request.content.iter_chunked(8192):
+            self._temp_file.write(chunk)
+            self._hasher.update(chunk)
+            self._size += len(chunk)
+            if self._size > self.max_size:
+                raise web.HTTPRequestEntityTooLarge(
+                    reason="File size exceeds the maximum limit.",
+                    max_size=self.max_size,
+                    actual_size=self._size,
+                )
+        self._temp_file.seek(0)
+        self._hash = self._hasher.hexdigest()
+
+    def __enter__(self):
+        self._temp_file.seek(0)
+        return self._temp_file
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._temp_file.close()
+        os.unlink(self._temp_file.name)
 
+    @property
+    def size(self) -> int:
+        return self._size
 
-class RawUploadedFile:
-    def __init__(self, content: Union[bytes, str]):
-        self.content = content
+    @property
+    def content(self) -> bytes:
+        self._temp_file.seek(0)
+        return self._temp_file.read()
 
     @property
-    def size(self) -> int:
-        return len(self.content)
+    def file(self):
+        return self._temp_file.name
+
+    def get_hash(self) -> str:
```

Comment:
Maybe this can move in the parent class? This is the same code on both classes

Reply:
And there is probably more code that can move to the parent class I think (but I haven't reviewed with this in mind)